### PR TITLE
New pymva keras callback: TensorBoard

### PR DESCRIPTION
**Pull request by pfackeldey**
This PR adds a new callback to the PyKeras method.
I've build ROOT without any errors and run the ctests and all of them passed. 

You can try out this new callback with:
`sudo pip install keras --upgrade`
`sudo pip install tensorflow --upgrade`

And then run: `python tutorials/tmva/keras/ClassificationKeras.py`. The callback will create a new directory called 'logs'.
Then you can call while or after training tensorboard: `tensorboard --logdir=./logs`
Now you can open the link from your terminal and visualize the network and training. 
The `ClassificationKeras.py` script, seems to be not compatible with the Keras 2.x API right now, therefore you can revert the changes of this PR in this specific script after testing.

More information on TensorBoard: https://www.tensorflow.org/get_started/summaries_and_tensorboard

Also pointing to the author of the PyKeras implementation: @stwunsch

Thanks in advance for reviewing this PR!

### CODE REVIEW

File: tutorials/tmva/keras/ClassificationKeras.py [hunk 1]
```diff
@@ -33,34 +32,38 @@
...
+                   '!H:!V:Fisher:VarTransform=D,G')
 factory.BookMethod(dataloader, TMVA.Types.kPyKeras, 'PyKeras',
-        'H:!V:VarTransform=D,G:FilenameModel=model.h5:NumEpochs=20:BatchSize=32')
+                   'H:!V:VarTransform=D,G:FilenameModel=model.h5:NumEpochs=20:BatchSize=32:Tensorboard=./logs')
 
 # Run training, test and evaluation
 factory.TrainAllMethods()
 factory.TestAllMethods()
-factory.EvaluateAllMethods()
+# factory.EvaluateAllMethods()
```
Comment (stwunsch) [inline tutorials/tmva/keras/ClassificationKeras.py]: Please remove this outcomment.

File: tutorials/tmva/keras/ClassificationKeras.py [hunk 2]
```diff
@@ -33,34 +32,38 @@
...
 model.save('model.h5')
 model.summary()
 
 # Book methods
 factory.BookMethod(dataloader, TMVA.Types.kFisher, 'Fisher',
-        '!H:!V:Fisher:VarTransform=D,G')
+                   '!H:!V:Fisher:VarTransform=D,G')
 factory.BookMethod(dataloader, TMVA.Types.kPyKeras, 'PyKeras',
-        'H:!V:VarTransform=D,G:FilenameModel=model.h5:NumEpochs=20:BatchSize=32')
+                   'H:!V:VarTransform=D,G:FilenameModel=model.h5:NumEpochs=20:BatchSize=32:Tensorboard=./logs')
```
Comment (stwunsch) [inline tutorials/tmva/keras/ClassificationKeras.py]: I would not set this as default in the tutorial's example. I assume that it fails if tensorboard is not installed and this would confuse most people.

**Comment by stwunsch** [comment]
Thanks for the PR! I want to test the code soon (tensorboard is awesome!), but here already two issues you should address:
1. Squash your commits into a single one (run `rebase -i HEAD~<number of commits` and squash them)
2. Fix the `clang-format` diff, go to the travis build (see above) and have a look at the job output. There should be a diff, which you can apply.

**Comment by pfackeldey** [comment]
Thanks for your comments :) I will update this PR as soon as possible!

**Comment by pfackeldey** [comment]
@stwunsch I fixed your 2 comments :)

**Comment by lmoneta** [comment]
Hi, 

 I think we can merge this PR. @stwunsch, are all the issues addressed ?

File: tutorials/tmva/keras/ClassificationKeras.py [hunk 3]
```diff
@@ -33,32 +32,36 @@
...
 dataloader.PrepareTrainingAndTestTree(TCut(''),
-        'nTrain_Signal=4000:nTrain_Background=4000:SplitMode=Random:NormMode=NumEvents:!V')
+                                      'nTrain_Signal=4000:nTrain_Background=4000:SplitMode=Random:NormMode=NumEvents:!V')
 
 # Generate model
 
 # Define initialization
+
+
 def normal(shape, name=None):
```
Comment (stwunsch) [inline tutorials/tmva/keras/ClassificationKeras.py]: Remove this definition of the custom init, you've removed it from the model (that's fine!). As well, please clean up the spacings around.

**Comment by stwunsch** [comment]
Hi! I've just tested the feature, really nice! But there's one problem on my system: You've activated the `write_images=False` option in the keras callback, but then I get the following error:

```
InternalError (see above for traceback): PNG encoding failed
```

I've compiled it without the enabled images, then everything runs fine.

@pfackeldey Do you have an idea why it's failing on my system with the enabled images? Haven't found something by google.

**Comment by stwunsch** [comment]
So the actual question: Is it something weird with my system (linux+arch) or is it a problem of the implementation?

**Comment by stwunsch** [comment]
Btw @lmoneta : That's how the tensorboard evaluation looks like:

blub

**Comment by pfackeldey** [comment]
@stwunsch Puh, I don't know... Are you using the latest version of TensorBoard? I have tested it on Ubuntu 16.04.

**Comment by pfackeldey** [comment]
PR is updated! I've removed the two empty lines :)

**Comment by stwunsch** [comment]
I'm running the newest version from pip, but it's something with a depedency. What are the images for?

**Comment by pfackeldey** [comment]
The model weights will be visualized in an image with grey scaling. I think the most useful case for this options is for image recognition with convolutional networks...

**Comment by stwunsch** [comment]
can you post the plot for the classification TMVA example?

**Comment by stwunsch** [comment]
because if it is not super useful, i would probably stick to the keras defaults:

```
keras.callbacks.TensorBoard(log_dir='./logs', histogram_freq=0, batch_size=32, write_graph=True, write_grads=False, write_images=False, embeddings_freq=0, embeddings_layer_names=None, embeddings_metadata=None)
```

**Comment by stwunsch** [comment]
As well, probably you want to feed the batch size from the option to the batch size in the callback.

**Comment by pfackeldey** [comment]
This is my output for "write_images=True". It is not really useful for anything except if you train on images directly.
write_images

**Comment by pfackeldey** [comment]
Also: I fixed your comment with the batch_size :)

**Comment by stwunsch** [comment]
what is the difference between `histogram_freq=0` and `...=1`? I would like to know the reason for the options that differ from the default. I would set the `write_images` to `False`, because the current PyKeras interface is not desigend for image data. What about `write_grads`?

**Comment by pfackeldey** [comment]
write_grads shows histograms of the development of the weights and biases and their gradients. These histograms are plotted in 3D in Tensorboard, where the 3rd axis is epochs. By that you can see a trend of the above mentioned quantities. histogram_freq = 1 means that these histograms are computed after each epoch; histogram_freq = 10 means that they are computed after 10 epochs... and so on. write_grads can not be used with histogram_freq = 0!

**Comment by pfackeldey** [comment]
I've updated the PR and changed the TensorBoard call arguments to the default ones from Keras except of "logdir" and "batch_size".

**Comment by stwunsch** [comment]
Alright, I think that is the best solution!

**Comment by stwunsch** [comment]
:+1: I've no more comments!

**Comment by pfackeldey** [comment]
Can it be merged now? :blush:

**Comment by stwunsch** [comment]
From my point of view, yes!

**Comment by etejedor** [comment]
@lmoneta @stwunsch could you double check that everything is indeed ok and merge?

**Comment by stwunsch** [comment]
I did double check it, works fine for me.

**Comment by etejedor** [comment]
@phsft-bot build

**Comment by pfackeldey** [comment]
Reading the console output of jenkins it seems, that always 100% of the tests successfully pass, but it fails copying a certain file afterwards. Is this a bug from my side? If so, any suggestions on how to fix it? :)

**Comment by stwunsch** [comment]
The problem is here: `java.io.FileNotFoundException: /tmp/log-parser_root-pullrequests-build_build_17889 (No space left on device)`

The node's disk space is full. It's not a problem of the PR, as far as I understand the output.

**Comment by etejedor** [comment]
Yes there is some space issue with the builds of the PRs today, will rebuild when that's fixed

**Comment by etejedor** [comment]
@phsft-bot build
